gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_send_plugin.rb

```diff
@@ -0,0 +1,82 @@
+require 'date'
+
+module Aws
+  module Plugins
+    class ClientMetricsSendPlugin < Seahorse::Client::Plugin
+
+      def add_handlers(handlers, config)
+        if config.client_side_monitoring
+          # AttemptHandler comes just before we would retry an error.
+          handlers.add(AttemptHandler, step: :sign, priority: 95)
+          # LatencyHandler is as close to sending as possible.
+          handlers.add(LatencyHandler, step: :sign, priority: 0)
+        end
+      end
+
+      class LatencyHandler < Seahorse::Client::Handler
+        def call(context)
+          start_time = Time.now
+          resp = @handler.call(context)
+          end_time = Time.now
+          latency = ((end_time - start_time) * 1000).to_i
+          context.metadata[:current_call_attempt].request_latency = latency
+          resp
+        end
+      end
+
+      class AttemptHandler < Seahorse::Client::Handler
+        def call(context)
+          request_metrics = context.metadata[:client_metrics]
+          attempt_opts = {
+            timestamp: DateTime.now.strftime('%Q').to_i,
+            fqdn: context.http_request.endpoint.host, # Is this always right?
+            region: context.config.region,
+            user_agent: context.http_request.headers["user-agent"],
+          }
+          # It will generally cause an error, but it is semantically valid for
+          # credentials to not exist.
+          if context.config.credentials
+            attempt_opts[:access_key] =
+              context.config.credentials.credentials.access_key_id
+            attempt_opts[:session_token] =
+              context.config.credentials.credentials.session_token
+          end
+          call_attempt = request_metrics.build_call_attempt(attempt_opts)
+          context.metadata[:current_call_attempt] = call_attempt
+
+          # Need to handle success/failure and storage in request_metrics here.
+          resp = @handler.call(context)
+          headers = context.http_response.headers
+          if headers.include?("x-amz-id-2")
+            call_attempt.x_amz_id_2 = headers["x-amz-id-2"]
+          end
+          if headers.include?("x-amz-request-id")
+            call_attempt.x_amz_request_id = headers["x-amz-request-id"]
+          end
+          if headers.include?("x-amzn-request-id")
+            call_attempt.x_amzn_request_id = headers["x-amzn-request-id"]
+          end
+          call_attempt.http_status_code = context.http_response.status_code
+          if e = resp.error
+            e_name = _extract_error_name(e)
+            e_msg = e.message
+            call_attempt.aws_exception_msg = "#{e_name}: #{e_msg}"
+          else
+            # Do I need anything here?
```

Comment:
This this comment something to be removed? Or for discussion?
I image for success requests, at this point, all data need have been collected already?

Reply:
I think so, I basically forgot to clear out some self-discussion comments before sending up the PR. Will clean up for second round.